JSX: TS1127 for <>#</>

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->
<!--
Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker.

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the CONTRIBUTING guidelines: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ
-->

<!-- If you have a QUESTION:
   THIS IS NOT A FORUM FOR QUESTIONS.
   Ask questions at http://stackoverflow.com/questions/tagged/typescript
    or https://gitter.im/Microsoft/TypeScript
-->

<!-- If you have a SUGGESTION:
  Most suggestion reports are duplicates, please search extra hard before logging a new suggestion.
  See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md
-->

<!-- If you have a BUG:
  Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  Version 2.8.0-dev.20180217

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** TS1127, jsx invalid character 

**Code**

```ts
const elm = (<>#</>);
```

**Expected behavior:**
Compile with no error.

**Actual behavior:**
`test.tsx(1,16): error TS1127: Invalid character.`

**Related Issues:** #4207

Thanks.

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -4176,8 +4176,9 @@ namespace ts {
             parseExpected(SyntaxKind.LessThanToken);
 
             if (token() === SyntaxKind.GreaterThanToken) {
-                parseExpected(SyntaxKind.GreaterThanToken);
+                // See below for explanation of scanJsxText
                 const node: JsxOpeningFragment = <JsxOpeningFragment>createNode(SyntaxKind.JsxOpeningFragment, fullStart);
+                scanJsxText();
                 return finishNode(node);
             }
 

--- a/tests/baselines/reference/tsxFragmentPreserveEmit.js
+++ b/tests/baselines/reference/tsxFragmentPreserveEmit.js
@@ -13,6 +13,7 @@ declare var React: any;
 <>hi</>; // text inside
 <><span>hi</span><div>bye</div></>; // children
 <><span>1</span><><span>2.1</span><span>2.2</span></><span>3</span></>; // nested fragments
+<>#</>; // # would cause scanning error if not in jsxtext
 
 //// [file.jsx]
 <></>; // no whitespace
@@ -21,3 +22,4 @@ declare var React: any;
 <>hi</>; // text inside
 <><span>hi</span><div>bye</div></>; // children
 <><span>1</span><><span>2.1</span><span>2.2</span></><span>3</span></>; // nested fragments
+<>#</>; // # would cause scanning error if not in jsxtext

--- a/tests/baselines/reference/tsxFragmentPreserveEmit.symbols
+++ b/tests/baselines/reference/tsxFragmentPreserveEmit.symbols
@@ -35,3 +35,4 @@ declare var React: any;
 >span : Symbol(JSX.IntrinsicElements, Decl(file.tsx, 1, 22))
 >span : Symbol(JSX.IntrinsicElements, Decl(file.tsx, 1, 22))
 
+<>#</>; // # would cause scanning error if not in jsxtext

--- a/tests/baselines/reference/tsxFragmentPreserveEmit.types
+++ b/tests/baselines/reference/tsxFragmentPreserveEmit.types
@@ -52,3 +52,6 @@ declare var React: any;
 >span : any
 >span : any
 
+<>#</>; // # would cause scanning error if not in jsxtext
+><>#</> : JSX.Element
+

--- a/tests/baselines/reference/tsxFragmentReactEmit.js
+++ b/tests/baselines/reference/tsxFragmentReactEmit.js
@@ -13,6 +13,7 @@ declare var React: any;
 <>hi</>; // text inside
 <><span>hi</span><div>bye</div></>; // children
 <><span>1</span><><span>2.1</span><span>2.2</span></><span>3</span></>; // nested fragments
+<>#</>; // # would cause scanning error if not in jsxtext
 
 //// [file.js]
 React.createElement(React.Fragment, null); // no whitespace
@@ -28,3 +29,4 @@ React.createElement(React.Fragment, null,
         React.createElement("span", null, "2.1"),
         React.createElement("span", null, "2.2")),
     React.createElement("span", null, "3")); // nested fragments
+React.createElement(React.Fragment, null, "#"); // # would cause scanning error if not in jsxtext

--- a/tests/baselines/reference/tsxFragmentReactEmit.symbols
+++ b/tests/baselines/reference/tsxFragmentReactEmit.symbols
@@ -35,3 +35,4 @@ declare var React: any;
 >span : Symbol(JSX.IntrinsicElements, Decl(file.tsx, 1, 22))
 >span : Symbol(JSX.IntrinsicElements, Decl(file.tsx, 1, 22))
 
+<>#</>; // # would cause scanning error if not in jsxtext

--- a/tests/baselines/reference/tsxFragmentReactEmit.types
+++ b/tests/baselines/reference/tsxFragmentReactEmit.types
@@ -52,3 +52,6 @@ declare var React: any;
 >span : any
 >span : any
 
+<>#</>; // # would cause scanning error if not in jsxtext
+><>#</> : JSX.Element
+

--- a/tests/cases/conformance/jsx/tsxFragmentPreserveEmit.tsx
+++ b/tests/cases/conformance/jsx/tsxFragmentPreserveEmit.tsx
@@ -15,3 +15,4 @@ declare var React: any;
 <>hi</>; // text inside
 <><span>hi</span><div>bye</div></>; // children
 <><span>1</span><><span>2.1</span><span>2.2</span></><span>3</span></>; // nested fragments
+<>#</>; // # would cause scanning error if not in jsxtext

--- a/tests/cases/conformance/jsx/tsxFragmentReactEmit.tsx
+++ b/tests/cases/conformance/jsx/tsxFragmentReactEmit.tsx
@@ -15,3 +15,4 @@ declare var React: any;
 <>hi</>; // text inside
 <><span>hi</span><div>bye</div></>; // children
 <><span>1</span><><span>2.1</span><span>2.2</span></><span>3</span></>; // nested fragments
+<>#</>; // # would cause scanning error if not in jsxtext

